### PR TITLE
add options.attachStacktrace

### PR DESCRIPTION
**Pull request by rxlabz**
## :loudspeaker: Type of change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring

## :scroll: Description

- add `SentryOptions.attachStackTrace` to attach stacktrace on captureMessage and captureEvent ( true by default )
- filter 'package:sentry' frames

## :bulb: Motivation and Context

https://develop.sentry.dev/sdk/unified-api/

## :green_heart: How did you test it?
new tests to verify if stacktrace is correctly sended

## :pencil: Checklist

- [ ] I reviewed submitted code
- [x] I added tests to verify changes
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

File: dart/lib/src/sentry_client.dart [hunk 1]
```diff
@@ -99,6 +100,15 @@ class SentryClient {
       event = event.copyWith(exception: sentryException);
     }
 
+    if (_options.attachStackTrace &&
```
Comment (rxlabz) [inline dart/lib/src/sentry_client.dart]: should we test if event.stackTrace is `null` ?

File: dart/lib/src/sentry_stack_trace_factory.dart [hunk 2]
```diff
@@ -101,14 +103,14 @@ class SentryStackTraceFactory {
 
     if (_inAppIncludes != null) {
       for (final include in _inAppIncludes) {
-        if (frame.package != null && frame.package.startsWith(include)) {
+        if (frame.package != null && frame.package == include) {
```
Comment (rxlabz) [inline dart/lib/src/sentry_stack_trace_factory.dart:108]: as we get a package name I'm not sure that `startWith` is the better option ?

File: dart/lib/src/protocol/sentry_event.dart [hunk 3]
```diff
@@ -262,6 +262,8 @@ class SentryEvent {
       json['exception'] = {
         'values': [exception.toJson()].toList(growable: false)
       };
+    } else if (stackTrace is SentryStackTrace) {
+      json['stacktrace'] = (stackTrace as SentryStackTrace).toJson();
```
Comment (marandaneto) [inline dart/lib/src/protocol/sentry_event.dart]: why this change? an exception has a stacktrace field already

Comment (rxlabz) [reply dart/lib/src/protocol/sentry_event.dart]: it's for case we want attach a stacktrace without exception ( for captureMessage and captureEvent )

Comment (marandaneto) [reply dart/lib/src/protocol/sentry_event.dart]: guess this would be necessary for captureMessage() as theres no exception

Comment (rxlabz) [reply dart/lib/src/protocol/sentry_event.dart]: ooops I just read that stacktrace should never be a top level property

Comment (marandaneto) [reply dart/lib/src/protocol/sentry_event.dart]: maybe we add one more field to the SentryEvent, because the docs about the stackTrace:

`Can be `null`, a [String], or a [StackTrace].`

it'd be confusing to add one more type, same as we have `throwable` and `exception`
wdyt?

File: dart/lib/src/sentry_client.dart [hunk 4]
```diff
@@ -99,6 +100,15 @@ class SentryClient {
...
+        event.throwable == null &&
+        event.exception == null) {
+      final frames =
+          SentryStackTraceFactory(_options).getStackFrames(StackTrace.current);
```
Comment (marandaneto) [inline dart/lib/src/sentry_client.dart]: please find a way to reuse the instance of `SentryStackTraceFactory`.

`SentryExceptionFactory` already creates an instance of it, it'd be easier to inject/test and also to not allocate memory on each call to captureEvent

File: dart/lib/src/sentry_stack_trace_factory.dart (hunk 2 above)
Comment (marandaneto) [reply dart/lib/src/sentry_stack_trace_factory.dart:108]: that's what I said before, on Java it makes sense because we have packages.like.this and inner packages belong to the parent package

**Comment by marandaneto** [comment]
missing this:

https://github.com/getsentry/sentry-dart/blob/main/dart/lib/src/sentry_exception_factory.dart#L30

this assignment should only be made if attachStacktrace is also enabled

File: dart/lib/src/protocol/sentry_event.dart (hunk 3 above)
Comment (marandaneto) [reply dart/lib/src/protocol/sentry_event.dart]: other solution would be to add the threads list

File: dart/lib/src/sentry_exception_factory.dart [hunk 5]
```diff
@@ -8,6 +8,8 @@ import 'sentry_stack_trace_factory.dart';
 class SentryExceptionFactory {
   SentryStackTraceFactory _stacktraceFactory;
 
+  SentryStackTraceFactory get stacktraceFactory => _stacktraceFactory;
```
Comment (marandaneto) [inline dart/lib/src/sentry_exception_factory.dart]: I'd not go this route, way better if SentryClient creates the instance and pass it along to SentryExceptionFactory, so both hold the reference and dont need to add a public getter, prefer DI sort of solution

File: dart/lib/src/sentry_client.dart (hunk 4 above)
Comment (rxlabz) [reply dart/lib/src/sentry_client.dart]: ✅

**Comment by rxlabz** [comment]
> missing this:
> 
> https://github.com/getsentry/sentry-dart/blob/main/dart/lib/src/sentry_exception_factory.dart#L30
> 
> this assignment should only be made if attachStacktrace is also enabled

✅

File: dart/lib/src/sentry_exception_factory.dart (hunk 5 above)
Comment (rxlabz) [reply dart/lib/src/sentry_exception_factory.dart]: ✅

File: dart/lib/src/sentry_client.dart [hunk 6]
```diff
@@ -1,6 +1,9 @@
 import 'dart:async';
 import 'dart:math';
 
+import 'package:meta/meta.dart';
+import 'package:sentry/src/sentry_stack_trace_factory.dart';
```
Comment (marandaneto) [inline dart/lib/src/sentry_client.dart]: avoid package:

File: dart/lib/src/protocol/sentry_event.dart (hunk 3 above)
Comment (rxlabz) [reply dart/lib/src/protocol/sentry_event.dart]: what about a `event.attachedStackTrace` property ? cause a `threads` interface only for a single property could also be confusing

File: dart/lib/src/sentry_client.dart [hunk 7]
```diff
@@ -20,21 +33,22 @@ class SentryClient {
...
-  final SentryOptions _options;
-
-  final Random _random;
-
-  final SentryExceptionFactory _exceptionFactory;
-
-  static final _sentryId = Future.value(SentryId.empty());
+    return SentryClient._(
+      options,
+      stackTraceFactory: SentryStackTraceFactory(options),
```
Comment (marandaneto) [inline dart/lib/src/sentry_client.dart]: why dont you do this inside of the private ctor like the `SentryExceptionFactory`? seems that each field is inited in a different place.

File: dart/lib/src/sentry_client.dart [hunk 8]
```diff
@@ -99,6 +113,14 @@ class SentryClient {
       event = event.copyWith(exception: sentryException);
     }
 
+    if (_options.attachStackTrace &&
+        event.throwable == null &&
+        event.exception == null) {
+      final frames = _stackTraceFactory.getStackFrames(StackTrace.current);
+
+      event = event.copyWith(stackTrace: SentryStackTrace(frames: frames));
```
Comment (marandaneto) [inline dart/lib/src/sentry_client.dart]: frames can be null or empty, we dont need to copyWIth if its the case

File: dart/lib/src/sentry_options.dart [hunk 9]
```diff
@@ -162,6 +162,14 @@ class SentryOptions {
     _sdk = sdk ?? _sdk;
   }
 
+  bool _attachStackTrace = true;
+
+  bool get attachStackTrace => _attachStackTrace;
```
Comment (marandaneto) [inline dart/lib/src/sentry_options.dart:176]: docs

File: dart/lib/src/sentry_client.dart (hunk 7 above)
Comment (rxlabz) [reply dart/lib/src/sentry_client.dart]: we need an "existing" instance of SentryStackTraceFactory to instantiate the SentryExceptionFactory

Comment (rxlabz) [reply dart/lib/src/sentry_client.dart]: we can't do 
```dart
SentryClient._(this._options)  : _stackTraceFactory = stackTraceFactory,
        _exceptionFactory = SentryExceptionFactory(
          options: _options,
          stacktraceFactory: _stackTraceFactory,
        ),
        _random = _options.sampleRate == null ? null : Random();
```

File: dart/lib/src/protocol/sentry_event.dart (hunk 3 above)
Comment (marandaneto) [reply dart/lib/src/protocol/sentry_event.dart]: this is just a toJson method, it should not have any business logic here

File: dart/lib/src/sentry_client.dart (hunk 8 above)
Comment (rxlabz) [reply dart/lib/src/sentry_client.dart]: ✅

File: dart/lib/src/sentry_client.dart (hunk 7 above)
Comment (marandaneto) [reply dart/lib/src/sentry_client.dart]: can we do both assignments then in the method body? make it non final if necessary, I think it'd be better

File: dart/lib/src/sentry_options.dart (hunk 9 above)
Comment (rxlabz) [reply dart/lib/src/sentry_options.dart:176]: ✅

File: dart/lib/src/sentry_client.dart (hunk 6 above)
Comment (rxlabz) [reply dart/lib/src/sentry_client.dart]: ✅

File: dart/lib/src/sentry_client.dart (hunk 1 above)
Comment (marandaneto) [reply dart/lib/src/sentry_client.dart]: yes only if its null, but actually I'd do this as `else if` with the exceptions check above, its either the exception with a stacktrace or a threads.stacktrace, not both, this way makes possible that we have a exception.stacktrace and still a threads.thread.stacktrace which is wrong.

Comment (rxlabz) [reply dart/lib/src/sentry_client.dart]: ✅

File: dart/lib/src/sentry_client.dart (hunk 7 above)
Comment (rxlabz) [reply dart/lib/src/sentry_client.dart]: ✅

File: dart/lib/src/protocol/sentry_event.dart (hunk 3 above)
Comment (rxlabz) [reply dart/lib/src/protocol/sentry_event.dart]: I meant adding a `attachedStackTrace ` field to SentryEvent
 
```dart 
if (exception != null) {
      json['exception'] = {
        'values': [exception.toJson()].toList(growable: false)
      };
} else if (attachedStackTrace != null) {
      json['threads'] = {
        'values': [
          {
            'id': 0,
            'stacktrace': (stackTrace as SentryStackTrace).toJson(),
          }
        ]
      };
    }
```
WDYT?

File: dart/lib/src/sentry_client.dart [hunk 10]
```diff
@@ -20,22 +31,19 @@ class SentryClient {
...
-  final SentryExceptionFactory _exceptionFactory;
-
-  static final _sentryId = Future.value(SentryId.empty());
-
   /// Instantiates a client using [SentryOptions]
-  SentryClient._(this._options, {SentryExceptionFactory exceptionFactory})
-      : _exceptionFactory =
-            exceptionFactory ?? SentryExceptionFactory(options: _options),
-        _random = _options.sampleRate == null ? null : Random();
+  SentryClient._(this._options)
```
Comment (marandaneto) [inline dart/lib/src/sentry_client.dart:39]: they (SentryStackTraceFactory, SentryExceptionFactory) should still be optional param for testing I guess? but only assign a value if they are null

File: dart/lib/src/sentry_client.dart [hunk 11]
```diff
@@ -97,6 +105,15 @@ class SentryClient {
           .getSentryException(event.throwable, stackTrace: event.stackTrace);
 
       event = event.copyWith(exception: sentryException);
+    } else if (_options.attachStackTrace &&
+        event.stackTrace == null &&
+        event.throwable == null &&
```
Comment (marandaneto) [inline dart/lib/src/sentry_client.dart]: at this point, `event.throwable` is not even used anymore, so not worth checking it

File: dart/lib/src/sentry_exception_factory.dart [hunk 12]
```diff
@@ -6,17 +6,21 @@ import 'sentry_stack_trace_factory.dart';
...
     @required SentryOptions options,
-  }) {
-    if (options == null) {
+  }) : _options = options {
+    if (_options == null) {
       throw ArgumentError('SentryOptions is required.');
     }
 
-    _stacktraceFactory = stacktraceFactory ?? SentryStackTraceFactory(options);
+    _stacktraceFactory = stacktraceFactory ?? SentryStackTraceFactory(_options);
```
Comment (marandaneto) [inline dart/lib/src/sentry_exception_factory.dart]: maybe `SentryStackTraceFactory` should be required too and throw ArgumentError if null, as SentryClient will take care of create a instance of SentryStackTraceFactory

File: dart/lib/src/sentry_client.dart (hunk 11 above)
Comment (rxlabz) [reply dart/lib/src/sentry_client.dart]: ✅

File: dart/lib/src/sentry_exception_factory.dart (hunk 12 above)
Comment (rxlabz) [reply dart/lib/src/sentry_exception_factory.dart]: ✅

File: dart/lib/src/sentry_client.dart (hunk 10 above)
Comment (rxlabz) [reply dart/lib/src/sentry_client.dart:39]: sorry but I'm not sure to understand, should we add factories param to SentryClient factory ctor and private ctor to allow users to pass their own factories ?

Comment (marandaneto) [reply dart/lib/src/sentry_client.dart:39]: no, the reasoning is not to allow custom factories, the reason is to make it easily testable, if you can unit test it like this (without fakes or mocks), its fine.

File: dart/lib/src/protocol/sentry_event.dart (hunk 3 above)
Comment (marandaneto) [reply dart/lib/src/protocol/sentry_event.dart]: we dont need `attachedStackTrace` to know if we should make the `threads` list or not, this could be done by checking if the exception is null or not.
by big concern here is `stackTrace as SentryStackTrace` as mentioned before.

Comment (rxlabz) [reply dart/lib/src/protocol/sentry_event.dart]: oh yes sorry I meant 
```dart
if (exception != null) {
      json['exception'] = {
        'values': [exception.toJson()].toList(growable: false)
      };
} else if (attachedStackTrace != null) {
      json['threads'] = {
        'values': [
          {
            'id': 0,
            'stacktrace': attachedStackTrace.toJson(),
          }
        ]
      };
    }
```

with `attachedStackTrace` being a SentryStackTrace.
Do you prefer the threads property options ?

File: dart/lib/src/sentry_client.dart (hunk 10 above)
Comment (rxlabz) [reply dart/lib/src/sentry_client.dart:39]: seems ok (for current tests)

File: dart/lib/src/protocol/sentry_event.dart (hunk 3 above)
Comment (marandaneto) [reply dart/lib/src/protocol/sentry_event.dart]: would be possible to make `stackTrace` strong typed as `SentryStackTrace`? and all the captureEvent, captureException etc takes the given stackTrace (from a try catch) as a melhod param, so we can pass the stackTrace along and we dont really need the stackTrace on the event being dynamic.

Comment (marandaneto) [reply dart/lib/src/protocol/sentry_event.dart]: so then your implementation would work and we would not need to call it `attachedStackTrace` which is confusing, but only stackTrace

File: dart/lib/src/protocol/sentry_event.dart [hunk 13]
```diff
@@ -262,6 +262,15 @@ class SentryEvent {
       json['exception'] = {
         'values': [exception.toJson()].toList(growable: false)
       };
+    } else if (stackTrace is SentryStackTrace) {
+      json['threads'] = {
+        'values': [
+          {
+            'id': 0,
+            'stacktrace': (stackTrace as SentryStackTrace).toJson(),
```
Comment (marandaneto) [inline dart/lib/src/protocol/sentry_event.dart]: wondering if its worth to have the thread list or at least a thread object for it, because the threads object has more fields and we could use the `Isolate.current` to populate them, example, thread name with `Isolate.current.debugName`, we can create a card for it, but not really necessary now

Comment (marandaneto) [inline dart/lib/src/protocol/sentry_event.dart]:
```suggestion
            'stacktrace': (stackTrace as SentryStackTrace).toJson(),
            'crashed': true,
            'name': 'Current Isolate',
```

File: dart/lib/src/protocol/sentry_event.dart [hunk 14]
```diff
@@ -262,12 +263,15 @@ class SentryEvent {
...
-    } else if (stackTrace is SentryStackTrace) {
+    }
+
+    if (stackTrace != null) {
       json['threads'] = {
         'values': [
           {
-            'id': 0,
-            'stacktrace': (stackTrace as SentryStackTrace).toJson(),
+            'id': Isolate.current.debugName,
```
Comment (marandaneto) [inline dart/lib/src/protocol/sentry_event.dart]: `debugName` is a String, `id` is an Int, why have you changed it?
threads have a `name` field

File: dart/lib/src/protocol/sentry_event.dart [hunk 15]
```diff
@@ -83,10 +85,9 @@ class SentryEvent {
   /// If this behavior is undesirable, consider using a custom formatted [message] instead.
   final dynamic throwable;
 
-  /// The stack trace corresponding to the thrown [exception].
-  ///
-  /// Can be `null`, a [String], or a [StackTrace].
-  final dynamic stackTrace;
+  /// The stack trace corresponding to the thrown [exception] or attachedStackTrace
+  /// see [SentryOptions.attachStackTrace]
```
Comment (marandaneto) [inline dart/lib/src/protocol/sentry_event.dart]: the docs are off, this is only used if the thrown `throwable` was null, the given `stackTrace` (method params) was also null and `attachStackTrace` is enabled

File: dart/lib/src/protocol/sentry_event.dart [hunk 16]
```diff
@@ -262,12 +263,15 @@ class SentryEvent {
       json['exception'] = {
         'values': [exception.toJson()].toList(growable: false)
       };
-    } else if (stackTrace is SentryStackTrace) {
+    }
+
+    if (stackTrace != null) {
```
Comment (marandaneto) [inline dart/lib/src/protocol/sentry_event.dart]: it was an `else if` and it was correct, not sure why you have changed it

Comment (rxlabz) [reply dart/lib/src/protocol/sentry_event.dart]: ✅

File: dart/lib/src/protocol/sentry_event.dart (hunk 15 above)
Comment (marandaneto) [reply dart/lib/src/protocol/sentry_event.dart]: basically its only used for captureMessage() which doesnt have an exception to set the stackTrace

File: dart/lib/src/sentry_client.dart [hunk 17]
```diff
@@ -97,6 +114,14 @@ class SentryClient {
           .getSentryException(event.throwable, stackTrace: event.stackTrace);
 
       event = event.copyWith(exception: sentryException);
+    } else if (_options.attachStackTrace &&
+        event.stackTrace == null &&
+        event.exception == null) {
+      final frames = _stackTraceFactory.getStackFrames(StackTrace.current);
```
Comment (marandaneto) [inline dart/lib/src/sentry_client.dart]: you should use the given `stackTrace` if not null and only if null and `attachStacktrace` is enabled, you use `StackTrace.current`

File: dart/lib/src/sentry_client.dart [hunk 18]
```diff
@@ -56,6 +65,14 @@ class SentryClient {
       _options.logger(SentryLevel.debug, 'No scope is defined');
     }
 
+    if (stackTrace != null) {
+      event = event.copyWith(
+        stackTrace: SentryStackTrace(
+          frames: _stackTraceFactory.getStackFrames(stackTrace),
+        ),
+      );
+    }
```
Comment (marandaneto) [inline dart/lib/src/sentry_client.dart]: this looks like being done twice, here and https://github.com/getsentry/sentry-dart/pull/160/files#r521245971

File: dart/lib/src/sentry_client.dart [hunk 19]
```diff
@@ -109,11 +134,17 @@ class SentryClient {
...
-      stackTrace: stackTrace,
       timestamp: _options.clock(),
     );
+    if (stackTrace != null) {
+      event = event.copyWith(
+        stackTrace: SentryStackTrace(
+          frames: _stackTraceFactory.getStackFrames(stackTrace),
+        ),
+      );
+    }
```
Comment (marandaneto) [inline dart/lib/src/sentry_client.dart]: why this again here?

File: dart/lib/src/sentry_exception_factory.dart [hunk 20]
```diff
@@ -6,39 +6,44 @@ import 'sentry_stack_trace_factory.dart';
...
-    _stacktraceFactory = stacktraceFactory ?? SentryStackTraceFactory(options);
+    if (_stacktraceFactory == null) {
+      throw ArgumentError('SentryStackTraceFactory is required.');
+    }
   }
 
   SentryException getSentryException(
     dynamic exception, {
-    dynamic stackTrace,
+    SentryStackTrace stackTrace,
```
Comment (marandaneto) [inline dart/lib/src/sentry_exception_factory.dart]: this should still be `dynamic` I believe, not entirely sure, you assume that `exception.stackTrace` is always populated and its not the case.

`exception.stackTrace` should only be set if theres no `exception/throwable` at all

File: dart/lib/src/protocol/sentry_event.dart (hunk 15 above)
Comment (marandaneto) [reply dart/lib/src/protocol/sentry_event.dart]: the docs for `attachStackTrace` is pretty clear about that btw

File: dart/lib/src/protocol/sentry_event.dart (hunk 14 above)
Comment (rxlabz) [reply dart/lib/src/protocol/sentry_event.dart]: ✅

File: dart/lib/src/sentry_client.dart (hunk 18 above)
Comment (rxlabz) [reply dart/lib/src/sentry_client.dart]: ✅

File: dart/lib/src/sentry_client.dart [hunk 21]
```diff
@@ -109,14 +101,18 @@ class SentryClient {
       platform: event.platform ?? sdkPlatform,
     );
 
-    if (event.throwable != null && event.exception == null) {
+    if (stackTrace != null && event.throwable == null) {
```
Comment (marandaneto) [inline dart/lib/src/sentry_client.dart]: you should test `event.exception == null` here too

File: dart/lib/src/sentry_client.dart (hunk 17 above)
Comment (rxlabz) [reply dart/lib/src/sentry_client.dart]: ✅

File: dart/lib/src/sentry_client.dart (hunk 21 above)
Comment (rxlabz) [reply dart/lib/src/sentry_client.dart]: ✅

File: dart/lib/src/sentry_exception_factory.dart [hunk 22]
```diff
@@ -6,34 +6,44 @@ import 'sentry_stack_trace_factory.dart';
...
+    if (stackTrace != null) {
+      sentryStackTrace ??= SentryStackTrace(
+        frames: _stacktraceFactory.getStackFrames(stackTrace),
+      );
+    } else if (exception is Error) {
+      sentryStackTrace ??= SentryStackTrace(
+        frames: _stacktraceFactory.getStackFrames(exception.stackTrace),
+      );
+    } else if (_options.attachStackTrace) {
+      sentryStackTrace ??= SentryStackTrace(
```
Comment (marandaneto) [inline dart/lib/src/sentry_exception_factory.dart]: why do we use `??=` if the var was just declared and its always null and `else if` branches

File: dart/lib/src/sentry_client.dart (hunk 17 above)
Comment (marandaneto) [reply dart/lib/src/sentry_client.dart]: wrong condition IMO, the `else if` branch was right, but it's missing this:

if (stackTrace == null) stackTrace = StackTrace.current
and:
final frames = _stackTraceFactory.getStackFrames(stackTrace);

File: dart/lib/src/sentry_client.dart [hunk 23]
```diff
@@ -92,11 +101,27 @@ class SentryClient {
...
 
       event = event.copyWith(exception: sentryException);
+    } else if (stackTrace == null &&
+        _options.attachStackTrace &&
+        event.exception == null) {
+      final frames = _stackTraceFactory.getStackFrames(StackTrace.current);
+
+      if (frames != null && frames.isNotEmpty) {
+        event = event.copyWith(stackTrace: SentryStackTrace(frames: frames));
+      }
```
Comment (marandaneto) [inline dart/lib/src/sentry_client.dart]: I suggest extracting this to a method on `_stackTraceFactory`, here and `getSentryException` method kinda does the same thing to convern a dynamic stackTrace to a SentryStackTrace and looks like both are different.

what we need is:

if theres an `exception`, do not set `event.stackTrace`, end.

if there's a `throwable`, do not set `event.stackTrace`, because the exception already has most probably an exception, end.

if theres not a `throwable` nor a `exception`, lets set the `event.stackTrace` only if `attachStacktrace` is enabled.

for all the cases above, if the `stackTrace` is not null, use it, if its null, use `stackTrace.current`

File: dart/lib/src/sentry_exception_factory.dart (hunk 22 above)
Comment (rxlabz) [reply dart/lib/src/sentry_exception_factory.dart]: ✅

File: dart/lib/src/sentry_client.dart (hunk 17 above)
Comment (rxlabz) [reply dart/lib/src/sentry_client.dart]: ✅

File: dart/lib/src/sentry_client.dart (hunk 23 above)
Comment (rxlabz) [reply dart/lib/src/sentry_client.dart]: yeah it was messy. I updated to 
```dart
    if (event.exception != null) return event;

    if (event.throwable != null) {
      final sentryException = _exceptionFactory
          .getSentryException(event.throwable, stackTrace: stackTrace);

      return event.copyWith(exception: sentryException);
    }

    if (stackTrace != null || _options.attachStackTrace) {
      stackTrace ??= StackTrace.current;
      final frames = _stackTraceFactory.getStackFrames(stackTrace);

      if (frames != null && frames.isNotEmpty) {
        event = event.copyWith(stackTrace: SentryStackTrace(frames: frames));
      }
    }

    return event;
```
I'm not sure we really need to create another method. WDYT ?

File: dart/lib/src/sentry_client.dart [hunk 24]
```diff
@@ -101,20 +101,16 @@ class SentryClient {
...
       final sentryException = _exceptionFactory
           .getSentryException(event.throwable, stackTrace: stackTrace);
 
-      event = event.copyWith(exception: sentryException);
-    } else if (_options.attachStackTrace && event.exception == null) {
+      return event.copyWith(exception: sentryException);
+    }
+
+    if (stackTrace != null || _options.attachStackTrace) {
       stackTrace ??= StackTrace.current;
```
Comment (marandaneto) [inline dart/lib/src/sentry_client.dart]: this condition seems wrong.

`stackTrace` can be null, and if null, get the StackTrace.current only if attachStackTrace is enabled

File: dart/lib/src/sentry_client.dart (hunk 23 above)
Comment (marandaneto) [reply dart/lib/src/sentry_client.dart]:
```
if (stackTrace != null || _options.attachStackTrace) {
      stackTrace ??= StackTrace.current;
```

still wrong.
the code looks better now, I'd still extract to a method like I said before, DRY

File: dart/lib/src/sentry_client.dart (hunk 24 above)
Comment (rxlabz) [reply dart/lib/src/sentry_client.dart]: sorry I don't understand, here :
- if stacktrace is null AND _options.attachStackTrace is true => we use StackTrace.current
- if stacktrace is not null AND _options.attachStackTrace is false =>  we use stacktrace
- if stacktrace is not null AND _options.attachStackTrace is true =>  we use stacktrace
- if stacktrace is null AND _options.attachStackTrace is false =>  we don't set event.stacktrace
which case is wrong ?

Comment (marandaneto) [reply dart/lib/src/sentry_client.dart]: actually its right, I read as `&&` instead of `||`, my bad sorry.
lets just be sure that we cover all the cases via unit tests

Comment (rxlabz) [reply dart/lib/src/sentry_client.dart]: ✅ added few tests to be sure

File: dart/lib/src/sentry_client.dart [hunk 25]
```diff
@@ -109,12 +129,17 @@ class SentryClient {
     Scope scope,
     dynamic hint,
   }) {
-    final event = SentryEvent(
+    var event = SentryEvent(
```
Comment (marandaneto) [inline dart/lib/src/sentry_client.dart]: does it need to be a `var`?

File: dart/lib/src/sentry_exception_factory.dart [hunk 26]
```diff
@@ -6,34 +6,44 @@ import 'sentry_stack_trace_factory.dart';
...
-    if (exception is Error) {
-      stackTrace ??= exception.stackTrace;
-    } else {
-      stackTrace ??= StackTrace.current;
+    SentryStackTrace sentryStackTrace;
+    if (stackTrace != null) {
+      sentryStackTrace = SentryStackTrace(
+        frames: _stacktraceFactory.getStackFrames(stackTrace),
+      );
+    } else if (exception is Error) {
```
Comment (marandaneto) [inline dart/lib/src/sentry_exception_factory.dart]: maybe we should check that `exception.stackTrace` is not null as well, can it be null?

**Review by marandaneto — APPROVED** [review]
LGTM

File: dart/lib/src/hub.dart [hunk 27]
```diff
@@ -73,6 +77,7 @@ class Hub {
         try {
           sentryId = await item.client.captureEvent(
             event,
+            stackTrace: stackTrace,
```
Comment (bruno-garcia) [inline dart/lib/src/hub.dart:80]: What about we make the stackTrace part of the hint?
The idea of hint was to have a single .. hint to deal with these extra things without having to expand the public API

Comment (marandaneto) [reply dart/lib/src/hub.dart:80]: the `hint` still could be used for other stuff as we do on Android, but stackTrace on dart is captured separately so it makes sense IMO to have this expanded API.
on Java, a stack trace is always part of the exception, which is not this case.
The other way would be adding a new field (that is also called stackTrace or sentryStackTrace) to the event, which is confusing.

File: dart/lib/src/sentry_exception_factory.dart (hunk 26 above)
Comment (rxlabz) [reply dart/lib/src/sentry_exception_factory.dart]: ✅ you right, it could
